pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+				}
+
+				instanceInfoClone, updated = recon.removeShareFromInstanceInfo(instanceInfoClone, shareName)
+			}
+		}
+
+		var err error
+		if updated {
+			klog.Infof("instanceInfo %q has updated share assignment, trying to update object", instanceInfoClone.Name)
+			instanceInfoClone, err = recon.updateInstanceInfoStatus(context.TODO(), instanceInfoClone)
+			if err != nil {
+				klog.Errorf("failed to update status subresource of instanceInfo %q: %v", instanceInfo.Name, err)
+				continue
+			}
+			instanceInfos[instanceURI] = instanceInfoClone
+		}
+	}
+
+	for _, shareInfo := range shareInfos {
+		if shareInfo.Status == nil || shareInfo.Status.InstanceHandle == "" {
+			scTag := shareInfo.Spec.InstancePoolTag
+			if scTag == "" {
+				klog.Errorf("shareInfo %q has empty instancePoolTag", shareInfo.Name)
+				continue
+			}
+
+			var instanceURI string
+			var err error
+			for _, instanceInfo := range instanceInfos {
+				if instanceFitShare(instanceInfo, shareInfo) {
+					instanceURI = util.IINameToInstanceURI(instanceInfo.Name)
+					instanceInfo, err = recon.assignInstanceInfo(instanceInfo, shareInfo.Name)
+					if err != nil {
+						klog.Errorf("failed to add share %q to instanceInfo %q: %v", shareInfo.Name, instanceInfo.Name, err)
+						continue
+					}
+					klog.Infof("share %q is now assigned to instance %q", shareInfo.Name, instanceURI)
+					instanceInfos[instanceURI] = instanceInfo
+					break
+				}
+			}
+			if err != nil {
+				continue
+			}
+
+			if instanceURI == "" {
+				klog.Infof("couldn't find instance to fit share %q, generating new instance", shareInfo.Name)
+
+				instanceURI, _ = file.GenerateMultishareInstanceURI(&file.MultishareInstance{
+					Project:  recon.cloud.Project,
+					Location: shareInfo.Spec.Region,
+					Name:     util.NewMultishareInstancePrefix + string(uuid.NewUUID()),
+				})
+
+				instanceInfo, err := recon.requestInstance(instanceURI, shareInfo.Spec.InstancePoolTag)
+				if err != nil {
+					klog.Errorf("failed to create new instanceInfo %q: %v", instanceURI, err)
+					continue
+				}
+				instanceInfo, err = recon.assignInstanceInfo(instanceInfo, shareInfo.Name)
+				if err != nil {
+					klog.Errorf("failed to add share %q to instanceInfo %q: %v", shareInfo.Name, instanceInfo.Name, err)
+					continue
+				}
+				klog.Infof("share %q is now assigned to instance %q", shareInfo.Name, instanceURI)
+				instanceInfos[instanceURI] = instanceInfo
+			}
+
+			shareInfo, err := recon.assignShareInfo(shareInfo, instanceURI)
+			if err != nil {
+				klog.Errorf("cannot update the instanceHandle of shareInfo %q: %v", shareInfo.Name, err)
+				continue
+			}
+			shareInfos[shareInfo.Name] = shareInfo
+		}
+	}
+
+	for instanceURI, instanceInfo := range instanceInfos {
+		if instanceInfo.DeletionTimestamp != nil {
+			continue
+		}
+
+		instanceInfoClone := instanceInfo.DeepCopy()
+		var updated bool
+		if instanceEmpty(instanceInfo) {
```

Comment:
This is a instance cleanup block. Should this be outside assignSharesToInstances() ?

Reply:
we certainly can move it out of this function but logically we should keep it inside the `assign` funciton.

the reason is that during the previous blocks there're modifications to the instance's share assignment on `instanceInfo.Status.ShareNames` but no resizing were attempted but they are waited to the very end here.

Reply:
Ack. Any ordering assumption please document correctly